File: socket-app/src/main/java/com/socket/handler/ClientHandler.java
package com.socket.handler;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import com.socket.utils.SessionManager;

public class ClientHandler implements Runnable {
    
    private final Socket clientSocket;
    private final SessionManager sessionManager;
    private final long sessionId;
    private String clientName;
    private AtomicInteger totalMessagesSent = new AtomicInteger(0);
    private AtomicLong totalBytesReceived = new AtomicLong(0);
    private Long sessionStartTime;

    public ClientHandler(Socket clientSocket, SessionManager sessionManager) {
        this.clientSocket = clientSocket;
        this.sessionManager = sessionManager;
        this.sessionId = UUID.randomUUID().getLeastSignificantBits();
        this.sessionStartTime = System.currentTimeMillis();
    }
    
    @Override
    public void run() {
        try {
            BufferedInputStream in = new BufferedInputStream(clientSocket.getInputStream());
            
            sessionManager.registerSession(sessionId, this);
            System.out.println("Total active sessions: " + sessionManager.getTotalActiveSessions());

            this.clientName = readMessage(in);

            String message;
            while ((message = readMessage(in)) != null) { 

                totalMessagesSent.incrementAndGet();
                totalBytesReceived.addAndGet(message.getBytes().length);

                System.out.println("Received message from username = [" + this.clientName + "], message = [" + message + "]");

                if ("quit".equalsIgnoreCase(message)) {
                    break;
                }
            }

        } catch (IOException e) {
            System.err.println("Exception in client handler: " + e.getMessage());
            e.printStackTrace();
        } finally {
            closeConnection();
        }
    }

    private String readMessage(BufferedInputStream in) throws IOException {
        byte[] buffer = new byte[2048];
        int bytesRead = in.read(buffer);
        if (bytesRead == -1) {
            return null; 
        }
        
        String encodedMessage = new String(buffer, 0, bytesRead);
        return encodedMessage;
    }

    private void closeConnection() {
        System.out.println("Closing client connection for username = " + this.clientName + "\nTotal messages sent: " + totalMessagesSent.get() + ", Total bytes received: " + totalBytesReceived.get() + ", Session duration: " + (System.currentTimeMillis() - sessionStartTime) + "ms");

        sessionManager.unregisterSession(sessionId);
        try {
            clientSocket.close();
        } catch (IOException e) {
            System.err.println("Error closing client socket: " + e.getMessage());
        }
    }
}

File: socket-app-python/handler/client_handler.py
import uuid
import time
from threading import Thread

class ClientHandler(Thread):

    def __init__(self, client_socket, session_manager):
        Thread.__init__(self)
        self.client_socket = client_socket
        self.session_manager = session_manager
        self.session_id = uuid.uuid4().int & (1<<64)-1  
        self.client_name = None
        self.total_messages_sent = 0
        self.total_bytes_received = 0
        self.session_start_time = time.time() * 1000  

    def run(self):
        try:
            with self.client_socket as in_stream:
                self.session_manager.register_session(self.session_id, self)

                print(f"Total active sessions: {self.session_manager.get_total_active_sessions()}")

                self.client_name = self.read_message(in_stream)

                while True:
                    message = self.read_message(in_stream)

                    if message is None:
                        break

                    self.total_messages_sent += 1
                    self.total_bytes_received += len(message.encode('utf-8'))

                    print(f"Received message from username = [{self.client_name}], message = [{message}]")

                    if message.lower() == "quit":
                        break

        except Exception as e:
            print(f"Exception in client handler: {e}")
        finally:
            self.close_connection()

    def read_message(self, in_stream):
        buffer = in_stream.recv(2048)
        if len(buffer) == 0:
            return None

        return buffer.decode('utf-8')

    def close_connection(self):
        print(f"Closing client connection for username = {self.client_name}\nTotal messages sent: {self.total_messages_sent}, Total bytes received: {self.total_bytes_received}, Session duration: {time.time() * 1000 - self.session_start_time}ms")

        self.session_manager.unregister_session(self.session_id)
        try:
            self.client_socket.close()
        except Exception as e:
            print(f"Error closing client socket: {e}")
